src/CodeGen_RS_Dev.cpp

```diff
@@ -0,0 +1,552 @@
+#include "BitWriter_3_2/ReaderWriter_3_2.h"
+#include "CodeGen_RS_Dev.h"
+#include "CodeGen_Internal.h"
+#include "IROperator.h"
+#include "IRPrinter.h"
+#include "Debug.h"
+#include "Target.h"
+#include "LLVM_Headers.h"
+#include "LLVM_Runtime_Linker.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::vector;
+using std::string;
+
+using namespace llvm;
+
+// Sniff the contents of a kernel to extracts the bounds of all the
+// thread indices (so we know how many threads to launch), and the
+// amount of shared memory to allocate.
+class ExtractBoundsNames : public IRVisitor {
+public:
+    string names[4];
+
+    ExtractBoundsNames() {
+        for (int i = 0; i < 4; i++) {
+            names[i] = "";
+        }
+    }
+
+private:
+    using IRVisitor::visit;
+
+    void visit(const For *op) {
+        if (CodeGen_GPU_Dev::is_gpu_var(op->name)) {
+            internal_assert(is_zero(op->min));
+        }
+
+        if (ends_with(op->name, ".__block_id_x")) {
+            names[0] = op->name;
+        } else if (ends_with(op->name, ".__block_id_y")) {
+            names[1] = op->name;
+        } else if (ends_with(op->name, ".__block_id_z")) {
+            names[2] = op->name;
+        } else if (ends_with(op->name, ".__block_id_w")) {
+            names[3] = op->name;
+        }
+        op->body.accept(this);
+    }
+};
+
+CodeGen_RS_Dev::CodeGen_RS_Dev(Target host) : CodeGen_LLVM(host) {
+    debug(2) << "Created CodeGen_RS_Dev for target " << host.to_string().c_str()
+             << "\n";
+#if !(WITH_RS)
+    user_error << "rs not enabled for this build of Halide.\n";
+#endif
+    context = new llvm::LLVMContext();
+}
+
+CodeGen_RS_Dev::~CodeGen_RS_Dev() { delete context; }
+
+void CodeGen_RS_Dev::add_kernel(Stmt stmt, const std::string &kernel_name,
+                                const std::vector<GPU_Argument> &args) {
+    internal_assert(module != NULL);
+
+    // Use [kernel_name] as the function name.
+    debug(2) << "In CodeGen_RS_Dev::add_kernel name=" << kernel_name << "\n";
+
+    ExtractBoundsNames bounds_names;
+    stmt.accept(&bounds_names);
+
+    // Type Definitions
+    StructType *StructTy_struct_rs_allocation =
+        module->getTypeByName("struct.rs_allocation");
+
+    // Put the arguments in the symbol table
+    vector<std::tuple<string, Value *> > globals_sym_names;
+
+    // Constant Definitions
+    ConstantAggregateZero *const_empty_allocation_struct =
+        ConstantAggregateZero::get(StructTy_struct_rs_allocation);
+    ConstantInt *const_0 =
+        ConstantInt::get(*context, APInt(32, StringRef("0"), 10));
+
+    NamedMDNode *rs_export_var =
```

Comment:
Halide isn't constrained to 80 columns; probably better to put this and similar lines on a single line